File: binaryTrees/36.cpp
#include "../common/tree.h"

using namespace Tree;

int solution(TreeNode* root, int& cont) {
    if (root == nullptr)
        return 0;

    int solLeft = solution(root->left, cont);
    int solRight = solution(root->right, cont);

    int tempSolLeft = 0;
    if (root->left && root->left->val == root->val)
        tempSolLeft = solLeft + 1;
    
    int tempSolRight = 0;
    if (root->right && root->right->val == root->val)
        tempSolRight = solRight + 1;
    
    cont = max(cont, tempSolLeft + tempSolRight);

    return max(tempSolLeft, tempSolRight);

}

int longestUnivaluePath(TreeNode* root) {
    int cont = 0;
    solution(root, cont);
    return cont;
}

int main() {
    TreeNode* root = createNode(5);
    root->left = createNode(4);
    root->right = createNode(5);
    root->left->left = createNode(1);
    root->left->right = createNode(1);
    root->right->right = createNode(1);

    cout << longestUnivaluePath(root);

    return 0;
}

File: binaryTrees/36.py
from typing import Optional, Any, List


class TreeNode:
    def __init__(self, val: int):
        self.val = val
        self.left: TreeNode = None
        self.right: TreeNode = None


def solution(root: TreeNode, cont: List[int]) -> int:
    if root is None:
        return 0

    sol_left = solution(root.left, cont)
    sol_right = solution(root.right, cont)

    sl = sol_left + 1 if root.left and root.left.val == root.val else 0
    sr = sol_right + 1 if root.right and root.right.val == root.val else 0

    cont[0] = max(cont[0], sr + sl)

    return max(sl, sr)


def longestUnivaluePath(root: Optional[TreeNode]) -> int:
    cont = [0]
    solution(root, cont)
    return cont[0]


if __name__ == "__main__":
    root = TreeNode(1)
    root.right = TreeNode(1)
    root.right.left = TreeNode(1)
    root.right.right = TreeNode(1)
    root.right.right.left = TreeNode(1)

    root.right.left.left = TreeNode(1)
    root.right.left.right = TreeNode(1)

    print(longestUnivaluePath(root))
